servicetalk-transport-netty-internal/src/main/java/io/servicetalk/transport/netty/internal/TransportObserverInitializer.java

```diff
@@ -16,46 +16,93 @@
 package io.servicetalk.transport.netty.internal;
 
 import io.servicetalk.transport.api.ConnectionObserver;
+import io.servicetalk.transport.api.ConnectionObserver.SecurityHandshakeObserver;
 import io.servicetalk.transport.api.TransportObserver;
 
 import io.netty.buffer.ByteBuf;
 import io.netty.buffer.ByteBufHolder;
 import io.netty.channel.Channel;
 import io.netty.channel.ChannelDuplexHandler;
+import io.netty.channel.ChannelHandler.Sharable;
 import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelPipeline;
 import io.netty.channel.ChannelPromise;
+import io.netty.handler.ssl.SslHandshakeCompletionEvent;
 
+import javax.net.ssl.SSLSession;
+
+import static io.servicetalk.transport.netty.internal.NettyPipelineSslUtils.extractSslSession;
+import static io.servicetalk.transport.netty.internal.TransportObserverUtils.assignConnectionError;
 import static io.servicetalk.transport.netty.internal.TransportObserverUtils.assignConnectionObserver;
+import static io.servicetalk.transport.netty.internal.TransportObserverUtils.securityHandshakeObserver;
 import static java.util.Objects.requireNonNull;
 
 /**
  * A {@link ChannelInitializer} that registers a {@link ConnectionObserver} for all channels.
  */
 public final class TransportObserverInitializer implements ChannelInitializer {
 
+    /**
+     * Tells which side is using secure connection.
+     */
+    public enum SecureSide {
+        CLIENT,
+        SERVER,
+        NONE
+    }
+
     private final TransportObserver transportObserver;
+    private final SecureSide secureSide;
 
     /**
      * Creates a new instance.
      *
      * @param transportObserver {@link TransportObserver} to initialize for the channel
+     * @param secureSide tells which side is using secure connection
      */
-    public TransportObserverInitializer(final TransportObserver transportObserver) {
+    public TransportObserverInitializer(final TransportObserver transportObserver, final SecureSide secureSide) {
         this.transportObserver = requireNonNull(transportObserver);
+        this.secureSide = secureSide;
     }
 
     @Override
     public void init(final Channel channel) {
         final ConnectionObserver observer = requireNonNull(transportObserver.onNewConnection());
         assignConnectionObserver(channel, observer);
-        channel.pipeline().addLast(new TransportObserverChannelHandler(observer));
+        final ChannelPipeline pipeline = channel.pipeline();
+        pipeline.addLast(new TransportObserverHandler(observer, secureSide));
     }
 
-    private static final class TransportObserverChannelHandler extends ChannelDuplexHandler {
+    private static final class TransportObserverHandler extends ChannelDuplexHandler {
         private final ConnectionObserver observer;
+        private final SecureSide secure;
+        private boolean handshakeStarted;
```

Comment:
rename to `handshakeStartNotified` to indicate that this does not reflect handshake start state but the notification to the observer.